File: my-folder/0042-trapping-rain-water/solution.ts
function trap(height: number[]): number {
    let result = 0;

    let l = 0, r = height.length - 1;

    let leftMax = height[l], rightMax = height[r]

    while (l < r) {
        if(leftMax < rightMax) {
            l++;
            leftMax = Math.max(leftMax, height[l]);
            result += leftMax - height[l];
        } else {
            r--;
            rightMax = Math.max(rightMax, height[r])
            result += rightMax - height[r]
        }
    }

    return result;
};

File: my-folder/0042-trapping-rain-water/solution.rb
# @param {Integer[]} height
# @return {Integer}
def trap(height)
    volume = 0
    left_max = 0
    right_max = 0

    l = 0
    r = height.length - 1
    i = 1

    while l < r do
        if(height[l] <= height[r])
            left_max = [left_max, height[l]].max
            volume += left_max - height[l]
            l += 1
        else
            right_max = [right_max, height[r]].max
            volume += right_max - height[r]
            r -= 1
        end
    end

    volume
end
